File: AoC_2023/10/2.cpp
#include <bits/stdc++.h>
#define ll long long
using namespace std;
int main(){
    freopen("test.txt", "r", stdin);

    ll n = 140;
    vector<string> a(n);
    for (ll i = 0; i < n; ++i){
        cin >> a[i];
    }
    ll m = a[0].size();

    ll start_i = 0;
    ll start_j = 0;
    for (ll i = 0; i < n; ++i){
        for (ll j = 0; j < m; ++j){
            if (a[i][j] == 'S'){
                start_i = i;
                start_j = j;
            }
        }
    }

    vector<vector<ll>> map_dist(n, vector<ll>(m, -1));
    //cout << "ok";
    vector<vector<ll>> used(n, vector<ll>(m, 0));
    set<vector<ll>> que;
    que.insert({0, start_i, start_j});
    ll ans = 0;

    while (!que.empty()){
        //cout << que.size();
        vector<ll> inp = *que.begin();
        ll w = inp[0];
        ll i = inp[1];
        ll j = inp[2];
        ans = max(ans, w);
        que.erase(que.begin());
        used[i][j] = 1;


        if (i > 0 && (a[i - 1][j] == '|' || a[i - 1][j] == 'F' || a[i - 1][j] == '7')){
            if (a[i][j] == 'S' || a[i][j] == 'J' || a[i][j] == 'L' || a[i][j] == '|'){
                if (!used[i - 1][j]) que.insert({w + 1, i - 1, j});
            }
        }

        if (i < n - 1 && (a[i + 1][j] == '|' || a[i + 1][j] == 'J' || a[i + 1][j] == 'L')){
            if (a[i][j] == 'S' || a[i][j] == 'F' || a[i][j] == '7' || a[i][j] == '|'){
                if (!used[i + 1][j]) que.insert({w + 1,i + 1, j});
            }
        }

        if (j > 0 && (a[i][j - 1] == '-' || a[i][j - 1] == 'F' || a[i][j - 1] == 'L')){
            if (a[i][j] == 'S' || a[i][j] == 'J' || a[i][j] == '7' || a[i][j] == '-'){
                if (!used[i][j - 1]) que.insert({w + 1,i, j - 1});
            }
        }

        if (j < m - 1 && (a[i][j + 1] == '-' || a[i][j + 1] == 'J' || a[i][j + 1] == '7')){
            if (a[i][j] == 'S' || a[i][j] == 'F' || a[i][j] == 'L' || a[i][j] == '-'){
                if (!used[i][j + 1]) que.insert({w + 1,i, j + 1});
            }
        }

        map_dist[i][j] = w;
    }

    cout << ans << "\n";



    /*
    for (ll i = 0; i < n; ++i){
        for (ll j = 0; j < m; ++j){
            cout << map_dist[i][j] << " ";
        }
        cout << "\n";
    }
    */


    a[start_i][start_j] = '7';
    ll cnt = 0;

    for (ll i = 0; i < n; ++i){
        ll balance = 0;
        for (ll j = 0; j < m; ++j){
            if (map_dist[i][j] != -1){
                if (a[i][j] == '|' || a[i][j] == 'J' || a[i][j] == 'L'){
                    balance = (balance + 1) % 2;
                }
            }else if (balance == 1){
                cnt++;
            }
        }
    }
    cout << cnt << "\n";




}

File: AoC_2024/11/2.py
f = open('test.txt')
a = list(map(int, f.read().strip().split(' ')))
data = [[0 for j in range(1000001)] for i in range(51)]
for j in range(1000001):
    data[0][j] = 1
for i in range(1, 51):
    for j in range(1000001):
        if j == 0:
            data[i][j] = data[i - 1][1]
        elif len(str(j)) % 2 == 0:
            m = len(str(j)) // 2
            k1 = int(str(j)[:m])
            k2 = int(str(j)[m:])
            data[i][j] = data[i - 1][k1] + data[i - 1][k2]
        elif j * 2024 < 1000001:
            data[i][j] = data[i - 1][j * 2024]
        elif i == 1:
            data[i][j] = 1
        else:
            j1 = j
            cnt = 1
            while len(str(j1)) % 2 != 0:
                j1 *= 2024
                cnt += 1
            m = len(str(j1)) // 2
            k1 = int(str(j1)[:m])
            k2 = int(str(j1)[m:])
            if i - cnt >= 0:
                data[i][j] = data[i - cnt][k1] + data[i - cnt][k2]
            else:
                data[i][j] = 1

for q in range(25):
    b = []
    for i in range(len(a)):
        if a[i] == 0:
            b.append(1)
        elif len(str(a[i])) % 2 == 0:
            m = len(str(a[i])) // 2
            k1 = int(str(a[i])[:m])
            k2 = int(str(a[i])[m:])
            b.append(k1)
            b.append(k2)
        else:
            b.append(a[i] * 2024)

    a = b
print(a)
ans = 0
for x in a:
    if len(str(x)) % 2 == 0:
        m = len(str(x)) // 2
        k1 = int(str(x)[:m])
        k2 = int(str(x)[m:])

        if len(str(k1)) % 2 == 0:
            m1 = len(str(k1)) // 2
            k3 = int(str(k1)[:m1])
            k4 = int(str(k1)[m1:])
            ans += data[48][k3] + data[48][k4]
        else:
            ans += data[49][k1]

        if len(str(k2)) % 2 == 0:
            m1 = len(str(k2)) // 2
            k3 = int(str(k2)[:m1])
            k4 = int(str(k2)[m1:])
            ans += data[48][k3] + data[48][k4]
        else:
            ans += data[49][k2]

    else:
        if x > 100000:
            cnt = 1
            while len(str(x)) % 2 != 0:
                x *= 2024
                cnt += 1
            m = len(str(x)) // 2
            k1 = int(str(x)[:m])
            k2 = int(str(x)[m:])
            ans += data[50 - cnt][k1] + data[50 - cnt][k2]
        else:
            ans += data[50][x]
print(ans)